grand_challenge_forge/partials/example-evaluation-method/example-evaluation-method{{cookiecutter._}}/helpers.py

```diff
@@ -139,26 +145,47 @@ def sigchld_handler(*_, **__):
                 except Exception as e:
                     handle_error(e, prediction=future_to_predictions[future])
         finally:
-            terminating_child_processes = True
-            _terminate_child_processes()
+            if not terminating_child_processes:
+                terminating_child_processes = True
+                _terminate_child_processes()
 
 
 def _terminate_child_processes():
-    current_process = psutil.Process(os.getpid())
-    children = current_process.children(recursive=True)
+    process = psutil.Process(os.getpid())
+    children = process.children(recursive=True)
     for child in children:
         try:
             child.terminate()
         except psutil.NoSuchProcess:
             pass  # Not a problem
 
     # Wait for processes to terminate
-    gone, still_alive = psutil.wait_procs(children, timeout=5)
+    _, still_alive = psutil.wait_procs(children, timeout=5)
 
     # Forcefully kill any remaining processes
     for p in still_alive:
-        print(f"Forcefully killing child process {p.pid}")
         try:
             p.kill()
         except psutil.NoSuchProcess:
             pass  # That is fine
+
+    _reap_all_children()
+
+
+def _reap_all_children():
+    """Reaps all child processes that have terminated"""
+    try:
+        while True:
+            pid, _ = os.waitpid(-1, os.WNOHANG)
+            if pid == 0:  # No more terminated child processes
```

Comment:
Are you assuming that this top level process is pid 0 in the container? Remember that this will be called as a subprocess of sagemaker shim, and/or will at least be called with docker init, so will not be pid 0.

Reply:
You are right, I think I copied this from somewhere where they had not taken being in a subprocess. I've updated it.